Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp

```diff
@@ -3955,10 +3955,18 @@ void MediaPlayerPrivateGStreamer::setVisibleInViewport(bool isVisible)
     if (!isVisible) {
         GstState currentState;
         gst_element_get_state(m_pipeline.get(), &currentState, nullptr, 0);
-        if (currentState > GST_STATE_NULL)
+        // WebKitMediaSrc can't properly handle PAUSED -> READY -> PAUSED, so we have to avoid transitioning
+        // back to READY when the player becomes visible.
+        GstState minimumState = isMediaSource() ? GST_STATE_PAUSED : GST_STATE_READY;
+        if (currentState >= minimumState)
             m_invisiblePlayerState = currentState;
         m_isVisibleInViewport = false;
-        gst_element_set_state(m_pipeline.get(), GST_STATE_PAUSED);
+        // Avoid setting the pipeline to PAUSED unless the playbin URL has already been set,
+        // otherwise it will fail, and may leave the pipeline stuck on READY with PAUSE pending.
+        if (!m_url.isValid())
+            return;
```

Comment:
Didn't the previous patch fix that by doing the load earlier? Would that make more sense than the check? What happens if the `return` branch is hit before the load?

Reply:
`MediaPlayerPrivateGStreamer::setVisibleInViewport` can also be called before the load in `MediaPlayer::loadWithNextMediaEngine`.

Reply:
Currently it will only be called if it's visible so it doesn't branch here, but that could change. Better guard for it.

Reply:
> What happens if the return branch is hit before the load?

Then it doesn't go into PAUSED, just remains in the current state. The subsequent `commitLoad()` does set it to PAUSED though after the URL was set.